File: oauth2-authorization-server-extended/authorization-server/src/main/java/com/onlyteo/sandbox/validator/PasswordConstraintValidator.java
package com.onlyteo.sandbox.validator;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import org.springframework.beans.BeanWrapperImpl;

/**
 * Constraint validator to check that two passwords are equal.
 */
public class PasswordConstraintValidator implements ConstraintValidator<ValidPassword, Object> {

    private String field;
    private String confirmField;
    private String message;

    @Override
    public void initialize(final ValidPassword annotation) {
        this.field = annotation.field();
        this.confirmField = annotation.confirmField();
        this.message = annotation.message();
    }

    @Override
    public boolean isValid(final Object value, final ConstraintValidatorContext context) {
        final var beanWrapper = new BeanWrapperImpl(value);
        final var fieldValue = beanWrapper.getPropertyValue(field);
        final var confirmFieldValue = beanWrapper.getPropertyValue(confirmField);
        if (fieldValue != null && !fieldValue.equals(confirmFieldValue)) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate(message)
                    .addPropertyNode(field)
                    .addConstraintViolation();
            context.buildConstraintViolationWithTemplate(message)
                    .addPropertyNode(confirmField)
                    .addConstraintViolation();
            return false;
        } else {
            return true;
        }
    }
}

File: spring-boot-oauth2-authorization-server-extended/spring-boot-oauth2-authorization-server-extended-server/src/main/kotlin/com/onlyteo/sandbox/validator/PasswordConstraintValidator.kt
package com.onlyteo.sandbox.validator

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import org.springframework.beans.BeanWrapperImpl

/**
 * Constraint validator to check that two passwords match.
 */
class PasswordConstraintValidator : ConstraintValidator<ValidPassword, Any> {

    private var field: String? = null
    private var confirmField: String? = null
    private var message: String? = null

    override fun initialize(annotation: ValidPassword) {
        this.field = annotation.field
        this.confirmField = annotation.confirmField
        this.message = annotation.message
    }

    override fun isValid(value: Any, context: ConstraintValidatorContext): Boolean {
        val beanWrapper = BeanWrapperImpl(value)
        val fieldValue = beanWrapper.getPropertyValue(field!!)
        val confirmFieldValue = beanWrapper.getPropertyValue(confirmField!!)!!
        if (fieldValue != null && fieldValue != confirmFieldValue) {
            context.disableDefaultConstraintViolation()
            context.buildConstraintViolationWithTemplate(message)
                .addPropertyNode(field)
                .addConstraintViolation()
            context.buildConstraintViolationWithTemplate(message)
                .addPropertyNode(confirmField)
                .addConstraintViolation()
            return false
        } else {
            return true
        }
    }
}